File: src/main/java/cc/mirukuneko/realtrainmodrenewed/model/MQOParser.java
package cc.mirukuneko.realtrainmodrenewed.model;

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class MQOParser {
    public static MQOModel parse(InputStream is, boolean compressed) throws IOException {
        if (compressed) {
            // .mqoz files are ZIP archives containing .mqo files
            ZipInputStream zis = new ZipInputStream(is);
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                String entryName = entry.getName();
                if (entryName != null && entryName.toLowerCase().endsWith(".mqo")) {
                    MQOModel model = parseMQO(zis);
                    zis.closeEntry();
                    // Don't close zis here, let the caller close the original InputStream
                    return model;
                }
                zis.closeEntry();
            }
            throw new IOException("No .mqo file found in .mqoz archive");
        } else {
            return parseMQO(is);
        }
    }

    private static MQOModel parseMQO(InputStream is) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8));
        MQOModel model = new MQOModel();
        
        RealTrainModRenewed.LOGGER.info("Starting MQO parsing");
        
        String line;
        int lineNum = 0;
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            lineNum++;
            if (lineNum < 50) {
                RealTrainModRenewed.LOGGER.info("Line {}: {}", lineNum, line);
            }
            if (line.startsWith("Material")) {
                RealTrainModRenewed.LOGGER.info("Found Material section");
                parseMaterials(reader, model);
            } else if (line.startsWith("Vertex")) {
                RealTrainModRenewed.LOGGER.info("Found Vertex section");
                parseVertices(reader, model);
            } else if (line.startsWith("Face")) {
                RealTrainModRenewed.LOGGER.info("Found Face section");
                parseFaces(reader, model);
            } else if (line.startsWith("Object")) {
                RealTrainModRenewed.LOGGER.info("Found Object section");
                parseObject(reader, model);
            }
        }
        
        RealTrainModRenewed.LOGGER.info("MQO parsing complete");
        reader.close();
        return model;
    }

    private static void parseMaterials(BufferedReader reader, MQOModel model) throws IOException {
        String line;
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (line.equals("}")) break;
            
            // MQO material format: mat_name(tex_file)
            // Also includes other properties like color, etc.
            int parenStart = line.indexOf('(');
            int parenEnd = line.indexOf(')');
            if (parenStart > 0 && parenEnd > parenStart) {
                String matName = line.substring(0, parenStart).trim();
                String texFile = line.substring(parenStart + 1, parenEnd).trim();
                if (!texFile.isEmpty()) {
                    net.minecraft.resources.Identifier texture = ModelLoader.resolveTexture(texFile);
                    model.addMaterial(matName, texture);
                }
            }
        }
    }

    private static void parseVertices(BufferedReader reader, MQOModel model) throws IOException {
        String line;
        int vertexCount = 0;
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (line.equals("}")) {
                RealTrainModRenewed.LOGGER.info("Vertex section ended, parsed {} vertices", vertexCount);
                break;
            }
            
            // Vertex format: vx vy vz
            String[] parts = line.split("\\s+");
            if (parts.length >= 3) {
                try {
                    float x = Float.parseFloat(parts[0]);
                    float y = Float.parseFloat(parts[1]);
                    float z = Float.parseFloat(parts[2]);
                    model.addVertex(x, y, z);
                    vertexCount++;
                } catch (NumberFormatException e) {
                    // Skip invalid lines
                }
            }
        }
    }

    private static void parseFaces(BufferedReader reader, MQOModel model) throws IOException {
        String line;
        int faceCount = 0;
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (line.equals("}")) {
                RealTrainModRenewed.LOGGER.info("Face section ended, parsed {} faces", faceCount);
                break;
            }
            
            // MQO face format: V0 V1 V2 M(mat_name) UV0 UV1 UV2
            // UV format: U(V_index) V(V_index)
            String[] parts = line.split("\\s+");
            List<Integer> vertexIndices = new ArrayList<>();
            List<float[]> uvList = new ArrayList<>();
            String materialName = "default";
            
            for (int i = 0; i < parts.length; i++) {
                String part = parts[i];
                if (part.startsWith("M(")) {
                    int end = part.indexOf(')');
                    if (end > 2) {
                        materialName = part.substring(2, end);
                    }
                } else if (part.startsWith("U(") || part.startsWith("V(")) {
                    // UV coordinate reference - skip for now, will parse separately
                } else {
                    try {
                        vertexIndices.add(Integer.parseInt(part));
                    } catch (NumberFormatException e) {
                        // Skip non-numeric parts
                    }
                }
            }
            
            // Parse UV coordinates from the line
            // MQO format: U(index) V(index) for each vertex
            for (int i = 0; i < parts.length; i++) {
                String part = parts[i];
                if (part.startsWith("U(") && i + 1 < parts.length && parts[i + 1].startsWith("V(")) {
                    try {
                        int uIndex = Integer.parseInt(part.substring(2, part.length() - 1));
                        int vIndex = Integer.parseInt(parts[i + 1].substring(2, parts[i + 1].length() - 1));
                        // For now, use indices as placeholder - proper UV parsing requires separate vertex array
                        uvList.add(new float[]{uIndex, vIndex});
                    } catch (NumberFormatException e) {
                        // Skip invalid UV
                    }
                }
            }
            
            if (!vertexIndices.isEmpty()) {
                int[] vi = vertexIndices.stream().mapToInt(i -> i).toArray();
                float[][] uvs = uvList.toArray(new float[0][]);
                model.addFace(vi, uvs, materialName);
                faceCount++;
            }
        }
    }

    private static void parseObject(BufferedReader reader, MQOModel model) throws IOException {
        String line = reader.readLine();
        if (line == null) return;
        
        line = line.trim();
        RealTrainModRenewed.LOGGER.info("parseObject called with line: {}", line);
        
        // MQO Object format: Object "name" { properties... vertex... face... }
        // The line after "Object" could be properties, not necessarily "{"
        // We need to find the opening brace or handle properties directly
        
        String name = "";
        if (line.startsWith("{")) {
            name = line.substring(1).trim();
        } else {
            // This line is a property, extract name from the Object line instead
            // The Object line was already consumed in the main loop
            name = "unknown";
        }
        
        model.addGroup(name);
        RealTrainModRenewed.LOGGER.info("Parsing Object: {}", name);
        
        // Parse nested sections within Object
        int lineCount = 0;
        int braceDepth = line.startsWith("{") ? 1 : 0;
        
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            lineCount++;
            if (lineCount < 10) {
                RealTrainModRenewed.LOGGER.info("Object line {}: {}", lineCount, line);
            }
            
            // Track brace depth
            if (line.equals("{")) {
                braceDepth++;
            } else if (line.equals("}")) {
                braceDepth--;
                if (braceDepth <= 0) {
                    RealTrainModRenewed.LOGGER.info("Object section ended");
                    break;
                }
            }
            
            // Match "vertex" or "face" followed by space and number, not "facet"
            if (line.matches("vertex\\s+\\d+\\s*\\{") || line.matches("Vertex\\s+\\d+\\s*\\{")) {
                RealTrainModRenewed.LOGGER.info("Found vertex section in object: {}", line);
                parseVertices(reader, model);
            } else if (line.matches("face\\s+\\d+\\s*\\{") || line.matches("Face\\s+\\d+\\s*\\{")) {
                RealTrainModRenewed.LOGGER.info("Found face section in object: {}", line);
                parseFaces(reader, model);
            }
        }
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/model/MQOParser.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.model

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream

object MQOParser {
    @JvmStatic
    @Throws(IOException::class)
    fun parse(inputStream: InputStream, compressed: Boolean): MQOModel {
        return if (compressed) {
            ZipInputStream(inputStream).use { zis ->
                var entry: ZipEntry? = zis.nextEntry
                while (entry != null) {
                    val entryName = entry.name
                    if (entryName != null && entryName.lowercase().endsWith(".mqo")) {
                        val model = parseMQO(zis)
                        zis.closeEntry()
                        return@use model
                    }
                    zis.closeEntry()
                    entry = zis.nextEntry
                }
                throw IOException("No .mqo file found in .mqoz archive")
            }
        } else {
            parseMQO(inputStream)
        }
    }

    @Throws(IOException::class)
    private fun parseMQO(inputStream: InputStream): MQOModel {
        val reader = BufferedReader(InputStreamReader(inputStream, StandardCharsets.UTF_8))
        val model = MQOModel()
        RealTrainModRenewed.LOGGER.info("Starting MQO parsing")

        var line: String?
        var lineNum = 0
        while (reader.readLine().also { line = it } != null) {
            var trimmed = line!!.trim()
            lineNum++
            if (lineNum < 50) RealTrainModRenewed.LOGGER.info("Line {}: {}", lineNum, trimmed)
            when {
                trimmed.startsWith("Material") -> {
                    RealTrainModRenewed.LOGGER.info("Found Material section")
                    parseMaterials(reader, model)
                }
                trimmed.startsWith("Vertex") -> {
                    RealTrainModRenewed.LOGGER.info("Found Vertex section")
                    parseVertices(reader, model)
                }
                trimmed.startsWith("Face") -> {
                    RealTrainModRenewed.LOGGER.info("Found Face section")
                    parseFaces(reader, model)
                }
                trimmed.startsWith("Object") -> {
                    RealTrainModRenewed.LOGGER.info("Found Object section")
                    parseObject(reader, model)
                }
            }
        }
        RealTrainModRenewed.LOGGER.info("MQO parsing complete")
        reader.close()
        return model
    }

    @Throws(IOException::class)
    private fun parseMaterials(reader: BufferedReader, model: MQOModel) {
        var line: String?
        while (reader.readLine().also { line = it } != null) {
            val trimmed = line!!.trim()
            if (trimmed == "}") break
            val parenStart = trimmed.indexOf('(')
            val parenEnd = trimmed.indexOf(')')
            if (parenStart > 0 && parenEnd > parenStart) {
                val matName = trimmed.substring(0, parenStart).trim()
                val texFile = trimmed.substring(parenStart + 1, parenEnd).trim()
                if (texFile.isNotEmpty()) {
                    val tex = ModelLoader.resolveTexture(texFile) ?: continue
                    model.addMaterial(matName, tex)
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun parseVertices(reader: BufferedReader, model: MQOModel) {
        var line: String?
        var vertexCount = 0
        while (reader.readLine().also { line = it } != null) {
            val trimmed = line!!.trim()
            if (trimmed == "}") {
                RealTrainModRenewed.LOGGER.info("Vertex section ended, parsed {} vertices", vertexCount)
                break
            }
            val parts = trimmed.split("\\s+".toRegex())
            if (parts.size >= 3) {
                try {
                    model.addVertex(parts[0].toFloat(), parts[1].toFloat(), parts[2].toFloat())
                    vertexCount++
                } catch (_: NumberFormatException) { }
            }
        }
    }

    @Throws(IOException::class)
    private fun parseFaces(reader: BufferedReader, model: MQOModel) {
        var line: String?
        var faceCount = 0
        while (reader.readLine().also { line = it } != null) {
            val trimmed = line!!.trim()
            if (trimmed == "}") {
                RealTrainModRenewed.LOGGER.info("Face section ended, parsed {} faces", faceCount)
                break
            }
            val parts = trimmed.split("\\s+".toRegex())
            val vertexIndices = mutableListOf<Int>()
            val uvList = mutableListOf<FloatArray>()
            var materialName = "default"

            for (part in parts) {
                when {
                    part.startsWith("M(") -> {
                        val end = part.indexOf(')')
                        if (end > 2) materialName = part.substring(2, end)
                    }
                    part.startsWith("U(") || part.startsWith("V(") -> { /* handled below */ }
                    else -> try { vertexIndices.add(part.toInt()) } catch (_: NumberFormatException) { }
                }
            }

            var i = 0
            while (i < parts.size) {
                val part = parts[i]
                if (part.startsWith("U(") && i + 1 < parts.size && parts[i + 1].startsWith("V(")) {
                    try {
                        val uIndex = part.substring(2, part.length - 1).toInt()
                        val vIndex = parts[i + 1].substring(2, parts[i + 1].length - 1).toInt()
                        uvList.add(floatArrayOf(uIndex.toFloat(), vIndex.toFloat()))
                    } catch (_: NumberFormatException) { }
                }
                i++
            }

            if (vertexIndices.isNotEmpty()) {
                model.addFace(vertexIndices.toIntArray(), uvList.toTypedArray(), materialName)
                faceCount++
            }
        }
    }

    @Throws(IOException::class)
    private fun parseObject(reader: BufferedReader, model: MQOModel) {
        var line = reader.readLine() ?: return
        line = line.trim()
        RealTrainModRenewed.LOGGER.info("parseObject called with line: {}", line)

        val name = if (line.startsWith("{")) line.substring(1).trim() else "unknown"
        model.addGroup(name)
        RealTrainModRenewed.LOGGER.info("Parsing Object: {}", name)

        var lineCount = 0
        var braceDepth = if (line.startsWith("{")) 1 else 0

        while (true) {
            val trimmed = reader.readLine()?.trim() ?: break
            lineCount++
            if (lineCount < 10) RealTrainModRenewed.LOGGER.info("Object line {}: {}", lineCount, trimmed)

            when (trimmed) {
                "{" -> braceDepth++
                "}" -> {
                    braceDepth--
                    if (braceDepth <= 0) {
                        RealTrainModRenewed.LOGGER.info("Object section ended")
                        break
                    }
                }
            }

            if (trimmed.matches(Regex("vertex\\s+\\d+\\s*\\{")) || trimmed.matches(Regex("Vertex\\s+\\d+\\s*\\{"))) {
                RealTrainModRenewed.LOGGER.info("Found vertex section in object: {}", trimmed)
                parseVertices(reader, model)
            } else if (trimmed.matches(Regex("face\\s+\\d+\\s*\\{")) || trimmed.matches(Regex("Face\\s+\\d+\\s*\\{"))) {
                RealTrainModRenewed.LOGGER.info("Found face section in object: {}", trimmed)
                parseFaces(reader, model)
            }
        }
    }
}
